### ISSUE TITLE
Line length guideline is drawn in front of text (instead of behind) in the script editor

### ISSUE THREAD

**Issue by Calinou** [godotengine/godot#56531]
### Godot version

4.0.dev (8f6cba3a4)

### System information

Fedora 34, GeForce GTX 1080 (NVIDIA 495.46)

### Issue description

The script editor's soft and hard line length guidelines are drawn in front of the text, while they should be drawn behind to avoid reducing readability.

![image](https://user-images.githubusercontent.com/180032/148290039-2be12cfb-ce51-4282-af1a-332e80f66a7a.png)

Full line:

![image](https://user-images.githubusercontent.com/180032/148290067-779ca364-48dd-40a9-93da-9ad1b898bb46.png)

In `3.x`, this is not the case since https://github.com/godotengine/godot/pull/30213 was merged. The line length guideline is drawn *behind* text as expected.

Fixing this in `master` should mostly involve changing the order of `draw_line()` calls in https://github.com/godotengine/godot/blob/3662105f8bd8c897caf625e44195bdfc7bdaf0f9/scene/gui/code_edit.cpp (search for `guideline`).

### Steps to reproduce

- Use the default code font and font size in the Editor Settings.
  - If using macOS, make sure to set the font hinting setting to **Light** instead of **Auto** (to get the same hinting as on Windows and Linux).
- Create a new scene with a Label node.
- Attach a new script to the Label node and paste the following contents:

```gdscript
extends Label


func _process(delta):
	text = "%d FPS (%.2 mspf)" % [Engine.get_frames_per_second, 1000.0 / Engine.get_frames_per_second()]

```

Notice how the `n` in the second `Engine.get_frames_per_second()` call is partially occluded by the hard line length guideline.

### Minimal reproduction project

N/A

**Comment by brennengreen** [comment]
From what I can tell we only use `draw_line()` once in `code_edit.cpp` for drawing underlines under function parameters.

We do some stuff with guidelines in:
```cpp
case NOTIFICATION_DRAW: {
			RID ci = get_canvas_item();
			const Size2 size = get_size();
			const bool caret_visible = is_caret_visible();
			const bool rtl = is_layout_rtl();
			const int row_height = get_line_height();

			if (line_length_guideline_columns.size() > 0) {
				const int xmargin_beg = style_normal->get_margin(SIDE_LEFT) + get_total_gutter_width();
				const int xmargin_end = size.width - style_normal->get_margin(SIDE_RIGHT) - (is_drawing_minimap() ? get_minimap_width() : 0);
				const int char_size = (int)font->get_char_size('0', 0, font_size).width;

				for (int i = 0; i < line_length_guideline_columns.size(); i++) {
					const int xoffset = xmargin_beg + char_size * (int)line_length_guideline_columns[i] - get_h_scroll();
					if (xoffset > xmargin_beg && xoffset < xmargin_end) {
						Color guideline_color = (i == 0) ? line_length_guideline_color : line_length_guideline_color * Color(1, 1, 1, 0.5);
						if (rtl) {
							RenderingServer::get_singleton()->canvas_item_add_line(ci, Point2(size.width - xoffset, 0), Point2(size.width - xoffset, size.height), guideline_color);
							continue;
						}
						RenderingServer::get_singleton()->canvas_item_add_line(ci, Point2(xoffset, 0), Point2(xoffset, size.height), guideline_color);
					}
				}
			}
```

Not entirely sure if that's related though

**Comment by Calinou** [comment]
@brennengreen Indeed, `canvas_item_add_line()` is a lower-level equivalent of `draw_line()`.

**Comment by Zylann** [comment]
Is 4.0 using a different font? The fact we uncover this problem seems to be that text gets rendered a bit more to the right than usual

**Comment by Calinou** [comment]
> Is 4.0 using a different font? The fact we uncover this problem seems to be that text gets rendered a bit more to the right than before

The font is the same as before, it's still Hack.

**Comment by mridsole** [comment]
I think this is due to the placement of the guidelines being incorrect in 4.0 (they're too far to the left by a proportional amount). Example with soft=20 and hard=30,

3.4: 
![guidelines-3 4](https://user-images.githubusercontent.com/9067729/148634404-89ee52bb-4b16-446f-8dc6-5b35096774a6.png)

4.0:
![guidelines-4 0](https://user-images.githubusercontent.com/9067729/148634414-a78a5c2d-64f6-4777-a90d-aabaa9df187d.png)

**Comment by NamanZelawat** [comment]
Hey @Calinou, I have replicated the issue but I am struggling with where specifically the call is made to draw both the lined present in the editor. Any suggestions will be helpful.

Thank you

**Comment by Calinou** [comment]
@NamanZelawat It seems this issue is actually caused by hinting setting mismatch:

> Edit: The actual issue seems to be that in 4.0, the text always renders as hinting=Normal, regardless of what it's set to, but the guidelines do adapt to the hinting setting. So with hinting=Normal it looks okay, but not with Light/None.

This is most likely not trivial to fix, so I removed the `junior job` label.

cc @bruvzg

**Comment by bruvzg** [comment]
Fix for guidelines #56787.

Hinting seems to be applied correctly, but for monospaced font, there's almost no difference in width (about 1 pixel at 100 character line).

**Comment by NamanZelawat** [comment]
> @NamanZelawat It seems this issue is actually caused by hinting setting mismatch:
> 
> > Edit: The actual issue seems to be that in 4.0, the text always renders as hinting=Normal, regardless of what it's set to, but the guidelines do adapt to the hinting setting. So with hinting=Normal it looks okay, but not with Light/None.
> 
> This is most likely not trivial to fix, so I removed the `junior job` label.
> 
> cc @bruvzg

Okay, @Calinou I didn't really understand what you are trying to convey but as you said it is not a junior job I will search for something else.

Thanks for replying.
